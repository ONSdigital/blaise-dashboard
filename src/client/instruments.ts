import { Instrument } from "blaise-api-node-client";
import axios from "axios";

export async function getInstruments(): Promise<Instrument[]> {
  const response = await axios.get("/api/instruments");

  return response.data;
}
