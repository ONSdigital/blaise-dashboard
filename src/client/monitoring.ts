import axios from "axios";
import { MonitoringDataModel } from "../server/monitoringDataModel";

export async function getMonitoring(): Promise<MonitoringDataModel[]> {
  const response = await axios.get("/api/monitoring");
  return response.data;
}
