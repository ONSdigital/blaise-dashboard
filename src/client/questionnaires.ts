import { Questionnaire } from "blaise-api-node-client";
import axios from "axios";

export async function getQuestionnaires(): Promise<Questionnaire[]> {
  const response = await axios.get("/api/questionnaires");

  return response.data;
}
