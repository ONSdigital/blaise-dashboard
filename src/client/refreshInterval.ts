export const refreshInterval = 60 * 1000; // 60 seconds
