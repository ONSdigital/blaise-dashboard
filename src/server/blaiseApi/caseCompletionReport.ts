import {calculateCaseTotals, calculatePercentComplete, CaseCalculations, CaseTotals} from "./calculations";
import {CaseStatus} from "blaise-api-node-client";


export type CaseCompletionReport = CaseTotals & CaseCalculations

export function buildCaseCompletionReport(cases: CaseStatus[]): CaseCompletionReport {
    const caseTotals = calculateCaseTotals(cases);
    const caseCompletePercentage = calculatePercentComplete(caseTotals)
    return {
        Total: caseTotals.Total,
        Complete: caseTotals.Complete,
        NotComplete: caseTotals.NotComplete,
        CompletePercentage: caseCompletePercentage
    }
}