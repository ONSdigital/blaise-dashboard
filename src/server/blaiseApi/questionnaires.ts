import BlaiseApiClient, { Questionnaire } from "blaise-api-node-client";
import { Config } from "../config";
import NodeCache from "node-cache";


export async function getQuestionnaires(
    blaiseApiClient: BlaiseApiClient,
    cache: NodeCache, config: Config,
    questionnaireTLA: string | undefined = undefined
): Promise<Questionnaire[]> {
    let listOfQuestionnaire: Questionnaire[] | undefined = cache.get("questionnaires");
    if (listOfQuestionnaire == undefined) {
        listOfQuestionnaire = await blaiseApiClient.getQuestionnaires(config.ServerPark);
        cache.set("questionnaires", listOfQuestionnaire);
    } else {
        console.log("hit cache for get questionnaires");
    }
    if (questionnaireTLA) {
        listOfQuestionnaire = filterQuestionnaires(listOfQuestionnaire, questionnaireTLA);
    }
    return listOfQuestionnaire;
}

export function filterQuestionnaires(listOfQuestionnaire: Questionnaire[], questionnaireTLA: string): Questionnaire[] {
    return listOfQuestionnaire.filter((questionnaire) => { return questionnaire.name.startsWith(questionnaireTLA); });
}
