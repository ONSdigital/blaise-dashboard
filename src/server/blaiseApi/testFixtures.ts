import {Outcome} from "blaise-api-node-client";

export const mockQuestionnaireList = [
        {
            name: "LMS2101_CC1",
            installDate: "210122",
            serverParkName: "gusty",
            fieldPeriod: "210122"
        },
        {
            name: "OPN2101A",
            installDate: "210122",
            serverParkName: "gusty",
            fieldPeriod: "210122"
        },
        {
            name: "OPN2007T",
            installDate: "210122",
            serverParkName: "gusty",
            fieldPeriod: "210122"
        }];

export const mockCaseList = [
            {
                primaryKey: "1000001" ,
                outcome: Outcome.Completed
            },
            {
                primaryKey: "1000002" ,
                outcome: Outcome.Completed
            },
            {
                primaryKey: "1000003" ,
                outcome: Outcome.None
            },
            {
                primaryKey: "1000004" ,
                outcome: Outcome.Partial
            },
            {
                primaryKey: "1000005" ,
                outcome: Outcome.AppointmentMade
            },
            {
                primaryKey: "1000006" ,
                outcome: Outcome.NonContact
            },
            {
                primaryKey: "1000007" ,
                outcome: 999
            }];