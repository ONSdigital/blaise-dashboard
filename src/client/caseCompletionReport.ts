import { CaseCompletionReport } from "../server/blaiseApi/caseCompletionReport";
import axios from "axios";

export async function getCaseCompletionReport(questionnaireName: string): Promise<CaseCompletionReport> {
  const response = await axios.get(`/api/reports/cases/completions/${questionnaireName}`);

  return response.data;
}
