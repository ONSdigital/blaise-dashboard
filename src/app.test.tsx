/**
 * @jest-environment jsdom
 */

import React from "react";
import { render, screen, waitFor, act } from "@testing-library/react";
import App from "./app";
import { mockQuestionnaireList } from "./server/blaiseApi/testFixtures";

jest.mock("./client/caseCompletionReport");
import { getCaseCompletionReport } from "./client/caseCompletionReport";
import { CaseCompletionReport } from "./server/blaiseApi/caseCompletionReport";

jest.mock("./client/questionnaires");
import { getQuestionnaires } from "./client/questionnaires";
import { Questionnaire } from "blaise-api-node-client";

const flushPromises = () => new Promise(setImmediate);

const getCaseCompletionReportMock = getCaseCompletionReport as jest.Mock<Promise<CaseCompletionReport>>;
const getQuestionnairesMock = getQuestionnaires as jest.Mock<Promise<Questionnaire[]>>;


describe("App", () => {
  const caseCompletionReport = {
    Total: 7,
    Complete: 5,
    NotComplete: 2,
    CompletePercentage: 71.43
  };

  afterEach(() => {
    jest.resetAllMocks();
  });

  it("renders correctly", async () => {
    getCaseCompletionReportMock.mockImplementation(() => Promise.resolve(caseCompletionReport));
    getQuestionnairesMock.mockImplementation(() => Promise.resolve(mockQuestionnaireList));

    const app = render(
      <App />
    );

    expect(screen.queryByText(/Loading/i)).toBeDefined();
    expect(screen.queryByText(/What is a completed case/i)).toBeDefined();

    await act(async () => {
      await flushPromises();
    });

    await waitFor(() => {
      expect(app).toMatchSnapshot();
    });
  });

  describe("when getting case completion reports errors", () => {
    it("renders an error panel", async () => {
      getQuestionnairesMock.mockImplementation(() => Promise.reject("Cannot get questionnaires"));

      render(
        <App />
      );

      await waitFor(() => {
        expect(screen.getByText("Failed to get questionnaires.")).toBeDefined();
        expect(screen.queryByText(/What is a completed case/i)).not.toBeUndefined();
      });
    });
  });


  describe("when no questionnaires are installed", () => {
    it("renders an info panel", async () => {
      getCaseCompletionReportMock.mockImplementation(() => Promise.resolve(caseCompletionReport));
      getQuestionnairesMock.mockImplementation(() => Promise.resolve([]));

      render(
        <App />
      );

      expect(screen.queryByText(/Loading/i)).toBeDefined();

      await waitFor(() => {
        expect(screen.getByText("No questionnaires installed.")).toBeDefined();
        expect(screen.queryByText(/What is a completed case/i)).not.toBeUndefined();
      });
    });
  });
});
