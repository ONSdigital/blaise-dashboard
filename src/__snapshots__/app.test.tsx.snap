// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="phase-banner"
      >
        <div
          class="container "
        >
          <div
            class="grid grid--flex grid--gutterless grid--vertical-center grid--no-wrap"
          >
            <div
              class="grid__col col-auto u-flex-no-grow"
            >
              <h3
                class="phase-banner__badge"
              >
                BETA
              </h3>
            </div>
            <div
              class="grid__col col-auto u-flex-shrink"
            >
              <p
                class="phase-banner__desc u-fs-s u-mb-no"
              >
                This is a new service – your 
                <a
                  href="https://ons.service-now.com/"
                >
                  feedback
                </a>
                 will help us improve it.
              </p>
            </div>
          </div>
        </div>
      </div>
      <header
        class="header header--internal"
      >
        <div
          class="header__top"
          role="banner"
        >
          <div
            class="container"
          >
            <div
              class="header__grid-top grid grid--gutterless grid--flex grid--between grid--vertical-center grid--no-wrap "
            >
              <div
                class="grid__col col-auto"
              >
                <a
                  class="header__logo-link"
                  href="/"
                >
                  <picture>
                    <img
                      alt="Office for National Statistics logo"
                      class="header__logo"
                      src="https://cdn.ons.gov.uk/sdc/design-system/31.4.0/img/ons-logo-neg-en.svg"
                    />
                  </picture>
                </a>
              </div>
            </div>
          </div>
        </div>
        <div
          class="header__main"
        >
          <div
            class="container"
          >
            <div
              class="grid grid--gutterless grid--flex grid--between grid--vertical-center grid--no-wrap"
            >
              <div
                class="grid__col col-auto u-flex-shrink"
              >
                <div
                  class="header__title"
                >
                  Dashboard
                </div>
              </div>
            </div>
          </div>
        </div>
      </header>
      <div
        class="page__container container"
        style="min-height: calc(67vh);"
      >
        <main
          class="page__main u-mt-no"
          id="main-content"
        >
          <h2
            class="u-mt-m"
          >
            Completed case information
          </h2>
          <table
            class="table"
            data-testid="case-report-table"
            id="case-report-table"
          >
            <thead
              class="table__head"
            >
              <tr
                class="table__row"
              >
                <th
                  class="table__header"
                  scope="col"
                >
                  Questionnaire
                </th>
                <th
                  class="table__header"
                  scope="col"
                >
                  Cases
                </th>
                <th
                  class="table__header"
                  scope="col"
                >
                  Completed
                </th>
                <th
                  class="table__header"
                  scope="col"
                >
                  Progress
                </th>
              </tr>
            </thead>
            <tbody
              class="table__body"
            >
              <tr
                class="table__row"
                data-testid="instrument-table-row"
              >
                <td
                  class="table__cell"
                  data-testid="instrument-case-report-questionnaire-LMS2101_CC1"
                >
                  LMS2101_CC1
                </td>
                <td
                  class="table__cell"
                  data-testid="instrument-case-report-total-LMS2101_CC1"
                >
                  7
                </td>
                <td
                  class="table__cell"
                  data-testid="instrument-case-report-complete-LMS2101_CC1"
                >
                  5
                   (
                  71.43
                  %)
                </td>
                <td
                  class="table__cell"
                  data-testid="instrument-case-report-complete-percentage-LMS2101_CC1"
                >
                  <progress
                    aria-valuemax="100"
                    aria-valuemin="0"
                    aria-valuenow="71.43"
                    id="file"
                    max="100"
                    role="progressbar"
                    value="71.43"
                  >
                    71.43
                    %
                  </progress>
                </td>
              </tr>
              <tr
                class="table__row"
                data-testid="instrument-table-row"
              >
                <td
                  class="table__cell"
                  data-testid="instrument-case-report-questionnaire-OPN2101A"
                >
                  OPN2101A
                </td>
                <td
                  class="table__cell"
                  data-testid="instrument-case-report-total-OPN2101A"
                >
                  7
                </td>
                <td
                  class="table__cell"
                  data-testid="instrument-case-report-complete-OPN2101A"
                >
                  5
                   (
                  71.43
                  %)
                </td>
                <td
                  class="table__cell"
                  data-testid="instrument-case-report-complete-percentage-OPN2101A"
                >
                  <progress
                    aria-valuemax="100"
                    aria-valuemin="0"
                    aria-valuenow="71.43"
                    id="file"
                    max="100"
                    role="progressbar"
                    value="71.43"
                  >
                    71.43
                    %
                  </progress>
                </td>
              </tr>
              <tr
                class="table__row"
                data-testid="instrument-table-row"
              >
                <td
                  class="table__cell"
                  data-testid="instrument-case-report-questionnaire-OPN2007T"
                >
                  OPN2007T
                </td>
                <td
                  class="table__cell"
                  data-testid="instrument-case-report-total-OPN2007T"
                >
                  7
                </td>
                <td
                  class="table__cell"
                  data-testid="instrument-case-report-complete-OPN2007T"
                >
                  5
                   (
                  71.43
                  %)
                </td>
                <td
                  class="table__cell"
                  data-testid="instrument-case-report-complete-percentage-OPN2007T"
                >
                  <progress
                    aria-valuemax="100"
                    aria-valuemin="0"
                    aria-valuenow="71.43"
                    id="file"
                    max="100"
                    role="progressbar"
                    value="71.43"
                  >
                    71.43
                    %
                  </progress>
                </td>
              </tr>
            </tbody>
          </table>
        </main>
      </div>
      <footer
        class="footer"
      >
        <div
          class="footer__body"
          data-analytics="footer"
        >
          <div
            class="container "
          >
            <div
              class="grid grid--flex grid--between"
            >
              <div
                class="grid__col"
              >
                <div
                  class="u-mt-m u-mb-s"
                >
                  <img
                    alt="Office for National Statistics"
                    src="https://ons-design-system.netlify.app/img/ons-logo-black-en.svg"
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
      </footer>
    </div>
  </body>,
  "container": <div>
    <div
      class="phase-banner"
    >
      <div
        class="container "
      >
        <div
          class="grid grid--flex grid--gutterless grid--vertical-center grid--no-wrap"
        >
          <div
            class="grid__col col-auto u-flex-no-grow"
          >
            <h3
              class="phase-banner__badge"
            >
              BETA
            </h3>
          </div>
          <div
            class="grid__col col-auto u-flex-shrink"
          >
            <p
              class="phase-banner__desc u-fs-s u-mb-no"
            >
              This is a new service – your 
              <a
                href="https://ons.service-now.com/"
              >
                feedback
              </a>
               will help us improve it.
            </p>
          </div>
        </div>
      </div>
    </div>
    <header
      class="header header--internal"
    >
      <div
        class="header__top"
        role="banner"
      >
        <div
          class="container"
        >
          <div
            class="header__grid-top grid grid--gutterless grid--flex grid--between grid--vertical-center grid--no-wrap "
          >
            <div
              class="grid__col col-auto"
            >
              <a
                class="header__logo-link"
                href="/"
              >
                <picture>
                  <img
                    alt="Office for National Statistics logo"
                    class="header__logo"
                    src="https://cdn.ons.gov.uk/sdc/design-system/31.4.0/img/ons-logo-neg-en.svg"
                  />
                </picture>
              </a>
            </div>
          </div>
        </div>
      </div>
      <div
        class="header__main"
      >
        <div
          class="container"
        >
          <div
            class="grid grid--gutterless grid--flex grid--between grid--vertical-center grid--no-wrap"
          >
            <div
              class="grid__col col-auto u-flex-shrink"
            >
              <div
                class="header__title"
              >
                Dashboard
              </div>
            </div>
          </div>
        </div>
      </div>
    </header>
    <div
      class="page__container container"
      style="min-height: calc(67vh);"
    >
      <main
        class="page__main u-mt-no"
        id="main-content"
      >
        <h2
          class="u-mt-m"
        >
          Completed case information
        </h2>
        <table
          class="table"
          data-testid="case-report-table"
          id="case-report-table"
        >
          <thead
            class="table__head"
          >
            <tr
              class="table__row"
            >
              <th
                class="table__header"
                scope="col"
              >
                Questionnaire
              </th>
              <th
                class="table__header"
                scope="col"
              >
                Cases
              </th>
              <th
                class="table__header"
                scope="col"
              >
                Completed
              </th>
              <th
                class="table__header"
                scope="col"
              >
                Progress
              </th>
            </tr>
          </thead>
          <tbody
            class="table__body"
          >
            <tr
              class="table__row"
              data-testid="instrument-table-row"
            >
              <td
                class="table__cell"
                data-testid="instrument-case-report-questionnaire-LMS2101_CC1"
              >
                LMS2101_CC1
              </td>
              <td
                class="table__cell"
                data-testid="instrument-case-report-total-LMS2101_CC1"
              >
                7
              </td>
              <td
                class="table__cell"
                data-testid="instrument-case-report-complete-LMS2101_CC1"
              >
                5
                 (
                71.43
                %)
              </td>
              <td
                class="table__cell"
                data-testid="instrument-case-report-complete-percentage-LMS2101_CC1"
              >
                <progress
                  aria-valuemax="100"
                  aria-valuemin="0"
                  aria-valuenow="71.43"
                  id="file"
                  max="100"
                  role="progressbar"
                  value="71.43"
                >
                  71.43
                  %
                </progress>
              </td>
            </tr>
            <tr
              class="table__row"
              data-testid="instrument-table-row"
            >
              <td
                class="table__cell"
                data-testid="instrument-case-report-questionnaire-OPN2101A"
              >
                OPN2101A
              </td>
              <td
                class="table__cell"
                data-testid="instrument-case-report-total-OPN2101A"
              >
                7
              </td>
              <td
                class="table__cell"
                data-testid="instrument-case-report-complete-OPN2101A"
              >
                5
                 (
                71.43
                %)
              </td>
              <td
                class="table__cell"
                data-testid="instrument-case-report-complete-percentage-OPN2101A"
              >
                <progress
                  aria-valuemax="100"
                  aria-valuemin="0"
                  aria-valuenow="71.43"
                  id="file"
                  max="100"
                  role="progressbar"
                  value="71.43"
                >
                  71.43
                  %
                </progress>
              </td>
            </tr>
            <tr
              class="table__row"
              data-testid="instrument-table-row"
            >
              <td
                class="table__cell"
                data-testid="instrument-case-report-questionnaire-OPN2007T"
              >
                OPN2007T
              </td>
              <td
                class="table__cell"
                data-testid="instrument-case-report-total-OPN2007T"
              >
                7
              </td>
              <td
                class="table__cell"
                data-testid="instrument-case-report-complete-OPN2007T"
              >
                5
                 (
                71.43
                %)
              </td>
              <td
                class="table__cell"
                data-testid="instrument-case-report-complete-percentage-OPN2007T"
              >
                <progress
                  aria-valuemax="100"
                  aria-valuemin="0"
                  aria-valuenow="71.43"
                  id="file"
                  max="100"
                  role="progressbar"
                  value="71.43"
                >
                  71.43
                  %
                </progress>
              </td>
            </tr>
          </tbody>
        </table>
      </main>
    </div>
    <footer
      class="footer"
    >
      <div
        class="footer__body"
        data-analytics="footer"
      >
        <div
          class="container "
        >
          <div
            class="grid grid--flex grid--between"
          >
            <div
              class="grid__col"
            >
              <div
                class="u-mt-m u-mb-s"
              >
                <img
                  alt="Office for National Statistics"
                  src="https://ons-design-system.netlify.app/img/ons-logo-black-en.svg"
                />
              </div>
            </div>
          </div>
        </div>
      </div>
    </footer>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
